### DISCUSSION TITLE
Performance of cocos2dx and axmol

### DISCUSSION THREAD

**Discussion by NgVThangBz** [axmolengine/axmol#1360] [General]
Has anyone tried to compare the performance of the two platforms? I can't say for sure, but according to my feelings, I noticed that cocos2dx 3.17 seems to consume less CPU at the same fps (RAM is equivalent).
i have a project, it used to run with cocos2dx 3.17, and i just migrated it to axmol, but i noticed the cpu in axmol is twice as high as cocos2dx, is this a bad sign,
Of course my project is not 100% completed yet, because some components have changed, has anyone done a CPU test yet?

**Comment by halx99** [comment]
refer to: https://github.com/axmolengine/axmol/issues/1279

**Comment by rh101** [comment]
> I can't say for sure, but according to my feelings, I noticed that cocos2dx 3.17 seems to consume less CPU at the same fps (RAM is equivalent).

That's not a fair comparison.  because Axmol is based on Cocos2d-x v4, which is significantly different to Cocos2d-x v3.17, so any comparison should be done between Axmol and Cocos2d-x v4.

Regardless of that, you should profile your app to see where the CPU is taking the most time, and if you find a problem or a possible case for optimization in the engine code, then create an issue request.

**Reply by NgVThangBz** [reply]
I tried with the sample project of cocos2dx 3.17, and axmol dev, (the 2 sample projects do not differ too much in code) but axmol consumes almost twice as much cpu as cocos3.17, I will also try with cocos4.0

**Reply by rh101** [reply]
I understand, but what I meant is that if you want valid comparison, then compare Cocos2d-x v3.17 to Cocos2d-x v4 to get a baseline for CPU usage, and then compare Cocos2d-x v4 to Axmol to see how it fairs.  Comparing Cocso2d-x v3.17 to Axmol doesn't make sense, given that Axmol was not based on that version.

**Comment by cigumo** [comment]
@NgVThangBz Which platform?

**Reply by NgVThangBz** [reply]
im use ios (16.6.1), ipad th7,
I only monitor the XCode debugger so can't confirm. and I'd like to ask for opinions from people who have done similar tests

**Comment by cigumo** [comment]
I've just run some tests for a production game on iOS 16.5 (iPhone XS Max), cocos 3.14, axmol 2.0.1.alpha1.

What I've noticed is running it from Xcode, there is 21% more CPU usage on Axmol (30% vs 51%), but I believe this is the Metal debugger, so I did another test. 

I ran a time profiler with the Release build. For the same scene as before, 1 second of profiling time spends 170ms in the cocos v3 version and 146ms in the axmol version. The drawScene calls take 145ms in cocos vs 125ms in axmol. 

So in this particular case I believe the Metal debugger is misleading and Axmol performance is actually better. 

Anyone did similar tests?

**Reply by rh101** [reply]
> I ran a time profiler with the Release build. For the same scene as before, 1 second of profiling time spends 170ms in the cocos v3 version and 146ms in the axmol version. The drawScene calls take 145ms in cocos vs 125ms in axmol.
> 
> So in this particular case I believe the Metal debugger is misleading and Axmol performance is actually better.

That is good to know!

Although, if you want to do a comparison between Cocos2d-x and Axmol, it really should be Cocos2d-x v4 against Axmol, because that is what Axmol is forked from.  Cocos2d-x v4 changed the rendering implementation, so it is very different from Cocos2d-x v3, and also uses Metal on iOS, whereas Cocos2d-x v3 only supports OpenGL for all platforms.

**Reply by aismann** [reply]
> So in this particular case I believe the Metal debugger is misleading and Axmol performance is actually better.

Axmol had some performance changes on spring this year or earlier.
